File: slides/number-theory/0_divisor_enum.py
def divisor(n: int) -> list[int]:
    assert(n >= 0)
    div = []
    x = 1
    while x * x <= n:
        if n % x == 0:
            div.append(x)
            if n != n // x:
                div.append(n // x)
        x += 1
    return div


if __name__ == '__main__':
    n = 12345
    for x in divisor(n):
        print(x)
    #> 1,12345,3,4115,5,2469,15,823

File: slides/number-theory/0_divisor_enum.cpp
#include <vector>
#include <cassert>
using namespace std;

vector<int> divisor(int n) {
    assert(n >= 0);
    std::vector<int> div;
    for (long long x = 1; x * x <= n; x++) {
        if (n % x == 0) {
            div.push_back(x);
            if (x != n / x) {
                div.push_back(n / x);
            }
        }
    }
    return div;
}


#include <iostream>

int main() {
    int n = 12345;
    for (int x: divisor(n)) cout << x << ',';
    //> 1,12345,3,4115,5,2469,15,823
}